src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetriesBehavior.cs

```diff
@@ -30,39 +43,51 @@ public SecondLevelRetriesBehavior(SecondLevelRetryPolicy retryPolicy, string loc
             }
             catch (Exception ex)
             {
-                var message = context.Message;
-                var currentRetry = GetNumberOfRetries(message.Headers) + 1;
+                failureInfoStorage.MarkForDeferralForSecondLevelRetry(context.Message.MessageId, ExceptionDispatchInfo.Capture(ex));
 
-                TimeSpan delay;
+                context.AbortReceiveOperation();
+            }
+        }
 
-                if (retryPolicy.TryGetDelay(message, ex, currentRetry, out delay))
-                {
-                    message.RevertToOriginalBodyIfNeeded();
-                    var messageToRetry = new OutgoingMessage(message.MessageId, message.Headers, message.Body);
+        async Task DeferMessageForSecondLevelRetry(ITransportReceiveContext context, Func<IRoutingContext, Task> fork, IncomingMessage message, ProcessingFailureInfo failureInfo)
+        {
+            var currentRetry = GetNumberOfRetries(message.Headers) + 1;
 
-                    messageToRetry.Headers[Headers.Retries] = currentRetry.ToString();
-                    messageToRetry.Headers[RetriesTimestamp] = DateTimeExtensions.ToWireFormattedString(DateTime.UtcNow);
+            var numberOfFirstLevelRetries = failureInfo.NumberOfFailedFlrAttempts - 1;
+            message.Headers[Headers.FLRetries] = numberOfFirstLevelRetries.ToString();
```

Comment:
@andreasohlund This means we now need to document that in v5 user can access all retry headers in their custom retry policies and in v6 they cannot. Is this what we want?